File: sdk/python/src/marmot/auth/workload/github.py
"""GitHub Actions OIDC source.

When a workflow has ``id-token: write`` permission, GitHub injects two env
vars and a runtime endpoint that vends an OIDC token. The token's audience
must match what Marmot is configured to accept.
"""

from __future__ import annotations

import os
from dataclasses import dataclass

import httpx

from marmot.auth.workload import SubjectToken


@dataclass
class GitHubActionsSource:
    name: str = "github-actions"
    audience: str | None = None  # defaults to the Marmot host at fetch time
    timeout: float = 5.0

    def fetch(self) -> SubjectToken | None:
        url = os.environ.get("ACTIONS_ID_TOKEN_REQUEST_URL")
        bearer = os.environ.get("ACTIONS_ID_TOKEN_REQUEST_TOKEN")
        if not url or not bearer:
            return None

        audience = self.audience or os.environ.get("MARMOT_HOST")
        params = {"audience": audience} if audience else {}

        try:
            resp = httpx.get(
                url,
                params=params,
                headers={"Authorization": f"Bearer {bearer}"},
                timeout=self.timeout,
            )
        except httpx.HTTPError:
            return None

        if resp.status_code != 200:
            return None

        try:
            token = resp.json().get("value")
        except ValueError:
            return None

        if not isinstance(token, str) or not token:
            return None
        return SubjectToken(token=token)

File: sdk/ts/src/auth/workload/github.ts
/** GitHub Actions OIDC source. */

import { TOKEN_TYPE_ID_TOKEN } from "../exchange.js";
import type { SubjectToken, WorkloadIdentitySource } from "./index.js";

export class GitHubActionsSource implements WorkloadIdentitySource {
  readonly name = "github-actions";
  private readonly audience: string | undefined;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: { audience?: string; timeoutMs?: number; fetchImpl?: typeof fetch } = {}) {
    this.audience = opts.audience;
    this.timeoutMs = opts.timeoutMs ?? 5_000;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  async fetch(): Promise<SubjectToken | null> {
    const env = processEnv();
    const url = env.ACTIONS_ID_TOKEN_REQUEST_URL;
    const bearer = env.ACTIONS_ID_TOKEN_REQUEST_TOKEN;
    if (!url || !bearer) return null;

    const audience = this.audience ?? env.MARMOT_HOST;
    const requestUrl = new URL(url);
    if (audience) requestUrl.searchParams.set("audience", audience);

    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), this.timeoutMs);
    try {
      const resp = await this.fetchImpl(requestUrl, {
        headers: { Authorization: `Bearer ${bearer}` },
        signal: ac.signal,
      });
      if (!resp.ok) return null;
      const json = (await resp.json()) as Record<string, unknown>;
      const value = json.value;
      if (typeof value !== "string" || !value) return null;
      return { token: value, tokenType: TOKEN_TYPE_ID_TOKEN };
    } catch {
      return null;
    } finally {
      clearTimeout(timer);
    }
  }
}

function processEnv(): Record<string, string | undefined> {
  return typeof process !== "undefined" && process.env ? process.env : {};
}
